src/NLog/Targets/DatabaseTarget.cs

```diff
@@ -712,28 +726,83 @@ protected string BuildConnectionString(LogEventInfo logEvent)
             sb.Append(";");
-            if (DBUserName == null)
+            var dbUserName = RenderLogEvent(DBUserName, logEvent);
+            if (string.IsNullOrEmpty(dbUserName))
             {
                 sb.Append("Trusted_Connection=SSPI;");
             }
             else
             {
                 sb.Append("User id=");
-                sb.Append(RenderLogEvent(DBUserName, logEvent));
+                sb.Append(dbUserName);
                 sb.Append(";Password=");
-                sb.Append(RenderLogEvent(DBPassword, logEvent));
+                var password = GetRenderedAndEscapedPassword(logEvent);
+                sb.Append(password);
                 sb.Append(";");
             }
-
-            if (DBDatabase != null)
+            var dbDatabase = RenderLogEvent(DBDatabase, logEvent);
+            if (!string.IsNullOrEmpty(dbDatabase))
             {
                 sb.Append("Database=");
-                sb.Append(RenderLogEvent(DBDatabase, logEvent));
+                sb.Append(dbDatabase);
             }
 
             return sb.ToString();
         }
 
+        private string GetRenderedAndEscapedPassword(LogEventInfo logEvent)
+        {
+            if (_fixedDbPassword != null)
+            {
+                return _fixedDbPassword;
+            }
+
+            var password = RenderLogEvent(DBPassword, logEvent);
+            password = EscapeValueForConnectionString(password);
+            return password;
+        }
+
+        /// <summary>
+        /// Escape quotes and semicolons.
+        /// See https://docs.microsoft.com/en-us/previous-versions/windows/desktop/ms722656(v=vs.85)#setting-values-that-use-reserved-characters
+        /// </summary>
+        private static string EscapeValueForConnectionString(string value)
+        {
+            const string singleQuote = "'";
+
+            if (value.StartsWith(singleQuote) && value.EndsWith(singleQuote))
+            {
+                // already escaped
+                return value;
+            }
+            const string doubleQuote = "\"";
+            if (value.StartsWith(doubleQuote) && value.EndsWith(doubleQuote))
+            {
+                // already escaped
+                return value;
+            }
+
+            var containsSingle = value.Contains(singleQuote);
+            var containsDouble = value.Contains(doubleQuote);
+            if (value.Contains(";") || containsSingle || containsDouble)
+            {
+                if (!containsSingle)
+                {
+                    return $"{singleQuote}{value}{singleQuote}";
+                }
+                if (!containsDouble)
+                {
+                    return $"{doubleQuote}{value}{doubleQuote}";
+                }
+
+                // both single and double
+                var escapedValue = value.Replace(doubleQuote, doubleQuote + doubleQuote);
+                return $"{doubleQuote}{escapedValue}{doubleQuote}";
```

Comment:
`string.Concat(doubleQuote, escapedValue, doubleQuote)` is faster.

Reply:
Any idea why the compiler rewrites it to `+`?

Reply:
What is rewritten to `+` ? This is string.Concat -> https://referencesource.microsoft.com/#mscorlib/system/string.cs,3216

Reply:
`$"{a}{b}"` -> `a+b` compile time? 

(anyway applied string.concat)

Reply:
seems to be string.format or string.concat by the compiler 👼 


> Compilation of interpolated strings
If an interpolated string has the type string, it's typically transformed into a String.Format method call. The compiler may replace String.Format with String.Concat if the analyzed behavior would be equivalent to concatenation.

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated

Reply:
$"{a}{b}" should become `string.Format("{0}{1}", a, b);`. Appends to StringBuilder from global-pool, and then performs ToString. `string.Concat` skips the StringBuilder-overhead.

Reply:
> seems to be string.format or string.concat by the compiler

Interesting. Wonder what version of the compiler this was introduced.

Reply:
Nevertheless, better to force it as join indeed!